File: 3360-minimum-deletions-to-make-string-k-special/minimum-deletions-to-make-string-k-special.go
func minimumDeletions(word string, k int) int {
	freq := make([]int, 26)

	// Đếm tần suất mỗi ký tự trong chuỗi
	for i := 0; i < len(word); i++ {
		freq[word[i]-'a']++
	}

	minDel := int(1e9)

	// Chọn mỗi ký tự làm "base", xem như chuẩn để so sánh
	for base := 0; base < 26; base++ {
		if freq[base] == 0 {
			continue
		}
		del := 0
		baseFreq := freq[base]

		for comp := 0; comp < 26; comp++ {
			if freq[comp] == 0 || comp == base {
				continue
			}

			compFreq := freq[comp]

			if compFreq < baseFreq {
				// Nếu nhỏ hơn base, xóa toàn bộ
				del += compFreq
			} else if compFreq-baseFreq > k {
				// Nếu lớn hơn base + k, xóa phần dư
				del += compFreq - baseFreq - k
			}
		}

		if del < minDel {
			minDel = del
		}
	}

	return minDel
}

File: 3360-minimum-deletions-to-make-string-k-special/minimum-deletions-to-make-string-k-special.rb
def minimum_deletions(word, k)
  freq = Array.new(26, 0)

  # Đếm tần suất mỗi ký tự
  word.each_char { |ch| freq[ch.ord - 'a'.ord] += 1 }

  min_del = Float::INFINITY

  # Thử mỗi ký tự làm chuẩn so sánh
  26.times do |base|
    next if freq[base] == 0

    del = 0
    base_freq = freq[base]

    26.times do |comp|
      next if freq[comp] == 0 || base == comp

      comp_freq = freq[comp]

      if comp_freq < base_freq
        del += comp_freq
      elsif comp_freq - base_freq > k
        del += comp_freq - base_freq - k
      end
    end

    min_del = [min_del, del].min
  end

  min_del
end
